channels/chan_simpleusb.c

```diff
@@ -344,6 +347,8 @@ static struct chan_simpleusb_pvt simpleusb_default = {
 	.rxondelay = 0,
 	.txoffdelay = 0,
 	.pager = PAGER_NONE,
+	.checkrxaudio = 1,
```

Comment:
I am wondering if this needs to be off until the hardware vendors begin including this in their hardware designs.

I understand that there is a plan to provide sample configurations for different vendor hardware.  This could be enabled in those samples for the hardware that supports this feature.

Reply:
There are a few reasons why I left the default as off initially for usbradio.
* I believe the majority of low-cost / small / radio-less nodes use simpleusb, whereas usbradio tends to be used more in repeaters or higher-end nodes ie. with commercial radios or mobile radios supporting flat audio. As a result, I suspect that people who are using usbradio are probably more likely on average to have properly calibrated their audio levels, and significantly less likely to be clipping their URI ADCs. So there should not be as much urgency to get this feature out to usbradio users, and, they are more probably likely to be more savvy about the config settings and can enable it easily if they want to. Also, usbradio users may be more likely to be using GPIOs for specialized purposes such as a repeater / remote base control line.
* There is also a serious issue in the existing code where there are some old "hacks" to match I/O levels to the original CM108 IC, which has been out of production for over 10 years (see "TBR" comments I added about this in the code), and thus in both simpleusb and usbradio these hacks are scaling the raw Tx audio by 110%, which **will** cause clipping in tx audio for all URIs that do not use the very old original CM108.  Those 3 TBR sections really should be removed now, but doing so will result in tx audio levels then being reduced by ~10% in both simpleusb and usbradio (for all URIs except those using the original CM108), and in rx levels being increased by ~2dB in usbradio (for all non-CM108 URIs). So for this reason I left the default as off for usbradio, with the idea that once these level matching hacks are removed and users made aware that they may need to adjust levels after upgrading, then at that point we could have checkrxaudio default to enabled.  And I would suggest that these hacks be removed ASAP, and could open a separate issue on that now if it would be helpful, and was going to do so either now or after this feature was added. (Seemed easier to add this first, then remove the old level scaling hacks as a separate issue and PR.)
* For simpleusb this feature should default to enabled because the users who are the most likely to be introducing clipped audio into the AllStarLink network are probably in general also those who are the least likely to later go in and enable the feature. In reviewing about a dozen URI schematics, none use GPIO1, and the very small number of simpleusb users who might use it as an output eg. for some sort of Event logic output, probably are familiar with the configuration settings and it should be easy for them to set the checkrxaudio parameter to off or to a different GPIO#.

Reply:
I think it would be better to default the GPIO pin usage to "off", period.  We don't know what hardware implementations are out in the wild and whether this new usage will trigger any conflicts.  It's a simple enough configuration change for those who do have the enhanced hardware.

Reply:
There are probably a relatively small number of users who are doing something with GPIO1, and if we wanted to be sure none of them could be inconvenienced by having the GPIO set if they are clipping their ADC, then yes the feature should default to off. In this case however I suspect that the benefit to AllStar as whole could be greater to enable the feature by default (since most users will just go with defaults and not bother to enable a feature if they have no immediate need to). So I would see it as a question of does ASL want to prioritize putting something in place that can over time help prevent clipping and thereby make a significant improvement in overall audio quality on the network, or, does ASL not want to do that in order to prevent any possible unexpected result to some very small number of users. To best make such a decision might require a bit of quantification of the benefits vs. potential inconveniences. I do believe that many 1,000's of users would greatly benefit over time from increased audio quality that results from AllStar users being more generally aware of clipping and audio levels, and I highly suspect that the number of simpleusb users who might have something unexpected happen after an upgrade (where they did not see the update notes about the checkrxaudio GPIO configuration setting) is probably relatively very small - probably < 0.1% of users. Thus from my experience I think the benefits would significantly outweigh the potential inconvenience. Also URI makers may be less likely to support a Clip LED if the feature is not enabled by default.

If other reviewers agree that the use of a GPIO should default to off I'll update the code, though I do think that this feature is a bit different than most in that it's intent is to protect the long-term audio quality of the ASL network as a whole, which is important and should be prioritized. In any other case I would not want to be the cause of a possible inconvenience to even a single user, but as it is now there many users who are inconvenienced on a daily basis when they have to ride their volume controls up and down on nets or larger QSOs due to one or two users having excessively loud/distorted. or overly quiet audio levels. There are a combination of approaches that could help address this over time, but I think that defining a Clip LED as a standard (and enabled by default) feature is a good start and a sort of acknowledgement that audio quality on ASL is a definite priority.

Reply:
Another potential suggestion... leave it disabled in the code and in configs/samples, but enable it for simpleusb in the configs/rpt version. In the code, it can stay disabled by default but the rpt sample config version could have it uncommented to enable by default.

Reply:
Q? isn't the clipping info available without enabling the GPIO pin?

Reply:
Re. is the clipping info available separately from the GPIO function, when I first implemented this I did not separate the stats collection and Clip LED output functionality but I then realized the stats data is very useful in the tune-menu and via AMI so I will update that now so that the statistics are always available.

And re. leaving disabled by default if the config is not present in the user's .conf file, but then enabling in the configs/rpt .conf files, if I understand that correctly it would result in it being enabled in new installs (where user will be reviewing and changing the various configs settings during node set up and thus will have a chance to review everything), but then not enabled by default for existing installs who do an ASL upgrade. That makes a lot of sense, I'll set that up now.

Reply:
Re: having the statistics always available ... YES!

Re: having the default GPIO pin usage (in code) set to disabled ... YES!

Re: enabling the GPIO pin usage by default in the configuration file ... I'm personally still hesitant here.  My preference would be to have the configuration file setting to enable/use one of the GPIO pins be commented out.  Those with hardware that supports the GPIO pin would only need to remove the ";".  and those who have hardware that uses that GPIO pin for something else won't run into conflicts.  But, that's "my" preference.

Reply:
> And re. leaving disabled by default if the config is not present in the user's .conf file, but then enabling in the configs/rpt .conf files, if I understand that correctly it would result in it being enabled in new installs (where user will be reviewing and changing the various configs settings during node set up and thus will have a chance to review everything), but then not enabled by default for existing installs who do an ASL upgrade. That makes a lot of sense, I'll set that up now.

Maybe I didn't state that super clearly... what I meant is that the configuration files in this repo go into configs/samples de facto, but there are also slightly different versions in configs/rpt "curated" for ASL.

The configs/samples files are not supposed to be ASL-catered at all, and hence I think the feedback on not trying to push users in a particular direction "for the sake of the network" would apply here. Same for the code, which isn't inherently ASL-specific either.

The configs in the rpt directory are intended for people who want an "ASL-recommended set of configs", and so I think it's fine to be a bit more hand holding there and set the default to 1 in that simpleusb.conf if you like, giving us the best of both worlds.

Reply:
I updated the code to not enable GPIO output by default in the .c files, so that after an ASL update GPIO writes will not result unless a user manually updates their .conf file to enable the clipledgpio parameter. For new installs the parameter is still set to 1 in simpleusb.conf. To clarify, on a new node when ASL3 is installed for the first time and it creates a new default .conf file, is that copied from the repository configs/rpt/simpleusb.conf?

Reply:
Might be a good question for @jxmx and the package maintainers... if just running the rpt_install.sh script, no, it uses the files in configs/samples only. But I think the ASL3 wrappers likely use the rpt files instead.

Reply:
> I think the ASL3 wrappers likely use the rpt files instead.

The default ASL3 configuration files come from .../app_rpt/configs/rpt/